### PR TITLE
[SPARK-33740][SQL] hadoop configs in hive-site.xml can overrides pre-existing hadoop ones

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?
org.apache.hadoop.conf.Configuration#setIfUnset will ignore those with defaults too

### Why are the changes needed?

fix a regression

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

new tests

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @cloud-fan @maropu  thanks

**Review by cloud-fan — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. I checked that the code is added by SPARK-31170 at 3.0.0, too.
Merged to master/3.1.

**Comment by dongjoon-hyun** [comment]
Could you make a backport to `branch-3.0` branch, @yaooqinn ?

File: sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala [hunk 1]
```diff
@@ -228,14 +228,21 @@ object SharedState extends Logging {
       sparkConf: SparkConf,
       hadoopConf: Configuration,
       initialConfigs: scala.collection.Map[String, String] = Map.empty): Unit = {
+
+    def containsInSparkConf(key: String): Boolean = {
+      sparkConf.contains(key) || sparkConf.contains("spark.hadoop." + key) ||
+        (key.startsWith("hive") && sparkConf.contains("spark." + key))
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala:234]: nit: It seems there is no test for the condition `(key.startsWith("hive") && sparkConf.contains("spark." + key))`?

File: sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala [hunk 2]
```diff
@@ -228,14 +228,21 @@ object SharedState extends Logging {
       sparkConf: SparkConf,
       hadoopConf: Configuration,
       initialConfigs: scala.collection.Map[String, String] = Map.empty): Unit = {
+
+    def containsInSparkConf(key: String): Boolean = {
+      sparkConf.contains(key) || sparkConf.contains("spark.hadoop." + key) ||
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala:233]: (This is not related to this PR though) The string prefix `"spark.hadoop."` seems to be used in many places, so how about defining a global value for it somewhere? I think that the refactoring could make it easy to track the positions where the prefix is used.

**Comment by maropu** [comment]
lgtm
